sim/src/component/userComponent.cpp

```diff
@@ -25,17 +32,54 @@ void UserComponent::setValue(const std::string& attrName,
             "Missing data type when setting value for attribute " + attrName);
     }
 
-    auto schemaType = compDef.schema().at(attrName);
+    // Ensure that data_type and value match
+    if (!proto::valHasCorrectDataType(value)) {
+        throw std::runtime_error(
+            "Data stored in value and data type stored in value do not match. "
+            "Data stored type: " +
+            proto::valStoredTypeName(value) +
+            ", data type: " + proto::valDataTypeName(value.data_type()));
+    }
+
+    auto& valToSet = values[attrName];
+    auto& schemaType = compDef.schema().at(attrName);
+    if (schemaType.has_array()) {
+        // Just set the array value completely
+        valToSet = value;
+        return;
+    }
+
+    // TODO(joeltio): Streamline this together with eqOp
+    // If schemaType and value are numeric, try to convert
+    if (proto::isTypeOfTypes<proto_NUMERIC>(schemaType) &&
+        proto::isValOfTypes<proto_NUMERIC>(value)) {
+        proto::runFnWithVal<proto_NUMERIC>(
+            value, [&valToSet, &schemaType]<class X>(X x) {
+                proto::runFnWithValType<proto_NUMERIC>(
+                    schemaType, [&valToSet, &x]<class Y>(Y* _) {
+                        proto::setVal(valToSet, (Y)x);
+                    });
+
+                // Return something random since it is required
+                return 3;
+            });
+        return;
+    }
+
+    // Otherwise, require that the value's data_type is the same as the
+    // schemaType
     auto differencer = google::protobuf::util::MessageDifferencer();
-    if (!differencer.Compare(schemaType, value.data_type())) {
+    // Ensure that the data_types between schema and value match
+    if (!differencer.Compare(value.data_type(), schemaType)) {
         throw std::runtime_error(
-            "Data type of given value does not match schema type for "
+            "Data type of given value does not match component's schemaType "
+            "for "
             "attribute " +
             attrName + ". Expected: " + proto::valDataTypeName(schemaType) +
             ", Got: " + proto::valDataTypeName(value.data_type()) + ".");
     }
 
-    values[attrName] = value;
+    valToSet = value;
```

Comment:
Is assign to a reference necessary here? I feel like its easier to read if you don't use references.

Reply:
It's not necessary, but I used it since I already defined it